ENH: Added DataFrame.compare and Series.compare (GH30429)

fujiaxiang opened the pull request:
Added `DataFrame.differences` and `Series.differences` methods. 

Have not yet added whatsnew entries. Will do so after review of API and behavior design.

A few design considerations open for discussion (among other things):
1. Index/column names: (self, other) vs (left, right) vs (old, new)
I used the (self, other) pair as I think they naturally match with the parameter names and the concept of comparing `self` with `other`. Although (left, right) or (old, new) may be slight more intuitive for users in some use cases, these new functions not necessarily compare two objects that are left and right (may be stacked on `axis=0`, or old and new.

1. Parameter name: axis
I feel axis may not be a very clear name, since we are stacking the output along this axis, rather than comparing objects along this axis. I considered using the name `stack_axis` or `orient` like `DataFrame.to_json` does.

1. I made the methods such that two NaNs are considered equal, hence not different. I think this should be the desired behavior since we are trying to compare two objects. However, under current implementation, a `None` will be considered equal to `np.nan` too. This is probably something we need to be careful for.

Let me know what you guys think. I'm ready to be scrutinized and criticized!

- [x] closes #30429
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

fujiaxiang committed:
ENH: Added DataFrame.differences and Series.differences (GH30429)

fujiaxiang committed:
CLN: reformatted docstring (GH30429)

WillAyd commented on pandas/core/frame.py:
```diff
@@ -5304,6 +5304,147 @@ def _construct_result(self, result) -> "DataFrame":
+        3    b     b  NaN   NaN  4.0   4.0
+        4    a     a  5.0   5.0  5.0   5.0
+        """
+        from pandas.core.reshape.concat import concat
```
Is this copy / paste for the Series method as well? If so should just place in NDFrame in pandas.core.generic

fujiaxiang replied:
No they are not. Both the docstring and the code are slightly different, although they share quite some common lines.

WillAyd replied:
Can you try to create a generic implementation in NDFrame? w.r.t. the docstring can use the Appender / Substitution decorators (will see other uses of them in pandas.core.generic)

fujiaxiang replied:
Have combined the two into NDFrame. 

I have kept the docstrings separate though, just like NDFrame.drop(), because their examples which take about two thirds of the space are completely different, and the rest are also not the same in that their input types are different, return types are different, and the explanations of arguments are also slightly different. 

I'm not sure whether it'd be good to decorate 2 Appenders thought, like maybe `Appender(generic._shared_docs["differences"])` followed by `Appender(_shared_docs["differences"])` (or maybe in reverse order?), where the former includes the input/output, then the latter includes the respective examples.

Let me know!

fujiaxiang committed:
ENH: Extracted differences() from DataFrame and Series into NDFrame

fujiaxiang committed:
Merge branch 'master' into dataframe_and_series_differences

fujiaxiang commented:
ping

WillAyd commented on pandas/core/frame.py:
```diff
@@ -5365,6 +5365,112 @@ def _construct_result(self, result) -> "DataFrame":
 
+    def differences(self, other, axis=1, keep_indices=False, keep_values=False):
+        """
+        Compare to another DataFrame and show the differences.
```
Can you place this docstring in the generic module and use the `Appender` / `Substitution` decorators to customize for the subclasses? Generally want to avoid copy / paste - will see other examples of how to do this in the module

fujiaxiang replied:
done

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

fujiaxiang committed:
ENH: organized docstring using _shared_doc and reduced duplicates (GH30429)

fujiaxiang committed:
ENH: added argument type indication (GH30429)

fujiaxiang committed:
ENH: reordered imports (GH30429)

fujiaxiang committed:
ENH: removed inconsistent type indication (GH30429)

jreback reviewed (changes requested):
will have a look soon.

cc @pandas-dev/pandas-core

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

# Conflicts:
#	pandas/core/series.py

fujiaxiang committed:
ENH: Added whatsnew entry (GH30429)

fujiaxiang committed:
ENH: Minor correction in whatsnew entry (GH30429)

fujiaxiang committed:
ENH: Minor correction in whatsnew entry (GH30429)

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

fujiaxiang committed:
ENH: Correction in whatsnew entry (GH30429)

fujiaxiang commented:
ping

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,31 @@ For example:
+       columns=["col1", "col2", "col3"],
+   )
+   df
+   df2 = df.copy()
```
can you split this into 2, by adding another ipython block, this is hard to see whats going on

fujiaxiang replied:
done, can you review?

jreback commented on pandas/core/frame.py:
```diff
@@ -5405,6 +5405,94 @@ def _construct_result(self, result) -> "DataFrame":
         out.columns = self.columns
         return out
 
+    @Appender(
```
is this added to the doc-string tests?

fujiaxiang replied:
I can't seem to find a doc-string test module, is there an example to refer to?

TomAugspurger replied:
See `ci/code_checks.sh`. Search for `--doctest-modules`.

fujiaxiang replied:
As only a selected few functions are tested in the existing set. I'm a bit confused here. What kind of functions/doc-strings need to be tested?

@TomAugspurger I tried adding the tests in the latest commit, could you take a look?

fujiaxiang replied:
resolving this conversation in favor of more updated discussion below

jreback commented on pandas/core/generic.py:
```diff
@@ -8180,6 +8180,90 @@ def ranker(data):
+        if axis in (1, "columns"):  # This is needed for Series
+            axis = 1
+        else:
+            axis = self._get_axis_number(axis)
```
doesn't self._get_axis_number(axis) just work?

fujiaxiang replied:
No, the issue is with `Series._get_axis_number` not allowing `axis=1` or `axis='columns'`.
We need to allow these 2 values in `Series.differences` because we are stacking the comparison result side by side

fujiaxiang replied:
makes sense. have renamed it to align_axis

jreback commented on pandas/core/generic.py:
```diff
@@ -8180,6 +8180,90 @@ def ranker(data):
+
+        diff = concat([self, other], axis=axis, keys=keys)
+
+        if axis >= self.ndim:
```
huh?

fujiaxiang replied:
Again, here we are handling `Series.differences(..., axis=1)`.
In general, we may always stack results in a new axis. For example, we may want to stack result of comparing two ``Series`` on `axis=1`, and in future cases we may want to stack result of comparing two ``DataFrame`` on `axis=2` etc.

jreback commented on pandas/core/generic.py:
```diff
@@ -8180,6 +8180,90 @@ def ranker(data):
+            return diff
+
+        ax = diff._get_axis(axis)
+        ax_names = np.array(ax.names)
```
why are you doing this?

fujiaxiang replied:
This is to deal with the same issue as #30485. I want to reorder the index levels according to their position, and not the index names. For example, if we stack the results on `axis='columns'`, we want to reorder the levels from `[0, 1]` to `[1, 0]` after we do concatenation with `keys=["self", "other"]`. In the rare that user dataframe has columns name `df.columns.name == [0]`, this would otherwise cause error.

That said, given this issue has affected 2 different places, we may want to refactor the `df.reorder_levels` method so we can indicate whether we want to reorder the positions explicitly, similar to what @TomAugspurger mentioned here #30289.

WillAyd replied:
I don't think this will work for MultiIndex at all

fujiaxiang replied:
@WillAyd It works. Actually this was written with MultiIndex in mind. 
Anyway I have added tests for MultiIndex cases. Can you review?

jreback reviewed (changes requested):
some comments, pls merge master

jreback commented:
cc @TomAugspurger @jorisvandenbossche

fujiaxiang committed:
ENH: updated whatsnew (GH31200)

fujiaxiang committed:
ENH: added doc references (GH31200)

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

# Conflicts:
#	pandas/core/series.py

fujiaxiang committed:
DOC: fixed formatting issue in doc references

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -8179,6 +8179,90 @@ def ranker(data):
+            * 1, or 'columns' : Stack differences on neighbouring columns.
+
+        keep_indices : bool, default False
+            Whether to keep the rows and columns that are equal, or drop them.
```
I don't really understand this, based just on the description. Can you think of a better description or keyword?

fujiaxiang replied:
I updated this to be `keep_shape` which should be clearer.

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -8179,6 +8179,90 @@ def ranker(data):
+        keep_indices : bool, default False
+            Whether to keep the rows and columns that are equal, or drop them.
+
+        keep_values : bool, default False
```
Should this be `keep_equal` rather than `keep_values`?

fujiaxiang replied:
updated

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -8179,6 +8179,90 @@ def ranker(data):
+            Whether to keep the rows and columns that are equal, or drop them.
+
+        keep_values : bool, default False
+            Whether to keep the values that are equal, or show as NaNs.
```
This reads like with `keep_values=False`, then equal values will be NaN. Is that correct?

fujiaxiang replied:
Yes that is correct. btw I've decided to change the name of this parameter to be `keep_equal` as per your earlier advice.

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -8179,6 +8179,90 @@ def ranker(data):
+            Object to compare with.
+
+        axis : {0 or 'index', 1 or 'columns'}, default 1
+            Determine how the differences are stacked.
```
I'm hopeful the examples will clear this up, but I'm not able to predict what the ouput will look like based just on this description.

fujiaxiang replied:
I updated the description a bit and hopefully this would clear things up

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -5401,6 +5401,94 @@ def _construct_result(self, result) -> "DataFrame":
+Returns
+-------
+DataFrame
+    DataFrame that shows the differences stacked side by side.
```
Probably worth stating a bit about what the index and columns will look like (multiindex or not, etc.)

fujiaxiang replied:
added

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -5401,6 +5401,94 @@ def _construct_result(self, result) -> "DataFrame":
+
+>>> df.differences(df2)
+  col1       col3
+  self other self other
```
Do we like `self` and `other`? Will users be familiar enough with `self`? Or should we use `left` and `right`?

fujiaxiang replied:
I mentioned this in the first message of this PR. 

I used the (self, other) pair as I think they naturally match with the parameter names and the concept of comparing `self` with `other`. Although in some cases (left, right) or (old, new) may be slight more intuitive for users, the key issue is that when user choose to stack results on `axis=0`, in which case the results are stacked vertically and `left-right` pair looks extremely weird. The `old-new` pair is also not ideal as sometimes users really are not comparing old data with new data, but simply comparing 2 different sets of stats.

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -5401,6 +5401,94 @@ def _construct_result(self, result) -> "DataFrame":
+
+Keep all the original indices (rows and columns)
+
+>>> df.differences(df2, keep_indices=True)
```
Why would this output be useful? Do we need two keywords for indices and values, or can we collapse them to one?

fujiaxiang replied:
I think this is useful in that users can decide on their own what they want to do about the cells that are equal. For example, they may want to replace the NaNs with `-` (hyphen). This is especially useful when you want to present multiple sets of differences so you see a consistent shape in results (e.g. `df1.differences(df2)` then `df3.differences(df4)`).

TomAugspurger reviewed (commented):
What happens if the two objects have different labels? Does this align prior to checking the differences? Will something like

```python
a = pd.Series([np.nan, np.nan])
b = pd.Series([])
a.differences(b)
```

have no differences?

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

fujiaxiang commented on pandas/core/generic.py:
```diff
@@ -8179,6 +8179,90 @@ def ranker(data):
+        The axis on which to stack results and how much information to
+        preserve can be customized.
+
+        Note that NaNs are considered not different from other NaNs.
```
done

fujiaxiang commented on pandas/core/generic.py:
```diff
@@ -8179,6 +8179,90 @@ def ranker(data):
+    ] = """
+        Compare to another %(klass)s and show the differences.
+
+        The axis on which to stack results and how much information to
```
done

fujiaxiang committed:
updated parameter names, docstring, and relevant tests (GH30429)

fujiaxiang committed:
added doc-string tests (GH30429)

pep8speaks commented:
Hello @fujiaxiang! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/frame.py`](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/frame.py):

> [Line 5153:63](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/frame.py#L5153): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 5154:89](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/frame.py#L5154): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 88 characters)
> [Line 5154:109](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/frame.py#L5154): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

* In the file [`pandas/core/generic.py`](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/generic.py):

> [Line 8163:75](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/generic.py#L8163): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 8165:79](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/generic.py#L8165): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 8169:89](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/generic.py#L8169): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 88 characters)
> [Line 8172:89](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/generic.py#L8172): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 88 characters)

* In the file [`pandas/core/series.py`](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/series.py):

> [Line 2582:57](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/series.py#L2582): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 2583:89](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/series.py#L2583): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 88 characters)
> [Line 2584:1](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/series.py#L2584): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 2585:62](https://github.com/pandas-dev/pandas/blob/586e37c56b3b49c0c110dd054aa5353d828fc47a/pandas/core/series.py#L2585): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

fujiaxiang commented:
> What happens if the two objects have different labels? Does this align prior to checking the differences? Will something like
> 
> ```python
> a = pd.Series([np.nan, np.nan])
> b = pd.Series([])
> a.differences(b)
> ```
> 
> have no differences?

This implementation can only compare `DataFrame` or `Series` that are identically-labelled. It raises the following error when this is violated:
```python
ValueError: Can only compare identically-labeled Series objects
```

The part that raises this error is the first step 
```python
self == other
```

fujiaxiang committed:
fixed some PEP8 issues in doc-strings (GH30429)

fujiaxiang committed:
removed trailing spaces in doc-strings (GH30429)

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

fujiaxiang committed:
fixed sphinx identation issues in doc-strings (GH30429)

fujiaxiang committed:
sphinx identation issues in doc-strings (GH30429)

fujiaxiang committed:
Merge branch 'master' into dataframe_and_series_differences

fujiaxiang committed:
sphinx identation issues in doc-strings (GH30429)

fujiaxiang commented:
Anyone can help me with the Sphinx doc-string warning? I tried many many times but I couldn't figure out what is wrong. Stackoverflow doesn't seem to have an existing answer either.

I keep getting this message:
```
/home/runner/work/pandas/pandas/pandas/core/frame.py:docstring of pandas.DataFrame.differences:13: WARNING: Unexpected indentation.
/home/runner/work/pandas/pandas/pandas/core/frame.py:docstring of pandas.DataFrame.differences:14: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/runner/work/pandas/pandas/pandas/core/series.py:docstring of pandas.Series.differences:13: WARNING: Unexpected indentation.
/home/runner/work/pandas/pandas/pandas/core/series.py:docstring of pandas.Series.differences:14: WARNING: Block quote ends without a blank line; unexpected unindent
```
then followed by 
```
build finished with problems, 4 warnings.
##[error]Process completed with exit code 1.
```
in the end.

WillAyd commented on pandas/core/frame.py:
```diff
@@ -5245,6 +5245,100 @@ def _construct_result(self, result) -> "DataFrame":
+
+See Also
+--------
+Series.differences: Show differences.
```
```suggestion
Series.differences : Show differences.
```

Technically need a space between these; this *might* be the CI issue you are seeing

WillAyd commented on pandas/core/series.py:
```diff
@@ -2574,7 +2574,77 @@ def _binop(self, other, func, level=None, fill_value=None):
+
+See Also
+--------
+DataFrame.differences: Show differences.
```
```suggestion
DataFrame.differences : Show differences.
```

fujiaxiang committed:
Update pandas/core/frame.py

minor changes in docstring

Co-Authored-By: William Ayd <william.ayd@icloud.com>

fujiaxiang committed:
Update pandas/core/series.py

minor changes in docstring

Co-Authored-By: William Ayd <william.ayd@icloud.com>

fujiaxiang commented:
@WillAyd thanks for the suggestion, let me try that

fujiaxiang committed:
attempt to fix sphinx identation issues in doc-strings (GH30429)

fujiaxiang committed:
removed trailing spaces in doc-strings (GH30429)

fujiaxiang committed:
Merge branch 'master' into dataframe_and_series_differences

fujiaxiang commented:
ping

WillAyd commented on ci/code_checks.sh:
```diff
@@ -259,17 +259,18 @@ fi
     MSG='Doctests frame.py' ; echo $MSG
-    pytest -q --doctest-modules pandas/core/frame.py
+    pytest -q --doctest-modules pandas/core/frame.py \
+        -k"-differences"
```
Is this excluding the differences test? Wasn't expecting a change like this

fujiaxiang replied:
@WillAyd no I didn't intend to exclude them from the tests. I wanted to include them in the tests but got a bit confused myself. Could you take a look again?

fujiaxiang committed:
removed unintended changes in ci/code_checks (GH30429)

fujiaxiang committed:
Merge remote-tracking branch 'origin/dataframe_and_series_differences' into dataframe_and_series_differences

fujiaxiang committed:
corrected errors in docstring (GH30429)

fujiaxiang commented:
ping

fujiaxiang committed:
Update pandas/core/frame.py: slight semantic cleanup in docstring

Co-Authored-By: William Ayd <william.ayd@icloud.com>

fujiaxiang committed:
renamed parameter axis to align_axis; added tests (GH30429)

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

fujiaxiang commented:
@WillAyd Your latest comment also appeared in the same thread as some of @jreback's earlier comments, so I replied there.
Basically I renamed the parameter `axis` to `align_axis` as I think your suggestion makes sense. I also added tests for ``MultiIndex" cases and the implementation works fine.

fujiaxiang committed:
minor correction in docstring

fujiaxiang committed:
some semantic cleanup in docstrings

jreback commented on pandas/core/frame.py:
```diff
@@ -5248,6 +5248,104 @@ def _construct_result(self, result) -> "DataFrame":
+    )
+    @Appender(_shared_docs["differences"] % _shared_doc_kwargs)
+    def differences(
+        self, other: "DataFrame", align_axis=1, keep_shape=False, keep_equal=False
```
can you type the args

fujiaxiang replied:
done

jreback commented on pandas/core/generic.py:
```diff
@@ -8268,6 +8268,92 @@ def ranker(data):
+        """
+
+    @Appender(_shared_docs["differences"] % _shared_doc_kwargs)
+    def differences(self, other, align_axis=1, keep_shape=False, keep_equal=False):
```
can you type the args

fujiaxiang replied:
done

jreback commented on pandas/core/series.py:
```diff
@@ -2577,7 +2577,81 @@ def _binop(self, other, func, level=None, fill_value=None):
+    )
+    @Appender(generic._shared_docs["differences"] % _shared_doc_kwargs)
+    def differences(
+        self, other: "Series", align_axis=1, keep_shape=False, keep_equal=False
```
can you type args

fujiaxiang replied:
done

jreback reviewed (changes requested):
what was the ratioale behind `align_axis`, we don't use this term anywhere else. what is wrong with `axis`?

WillAyd commented:
> what was the ratioale behind `align_axis`, we don't use this term anywhere else. what is wrong with `axis`?

My thought was that `axis` typically refers to how a function is applied against a frame. In this case things are applied element wise regardless and then `axis` or `align_axis` is used to determine how the construct the output

jorisvandenbossche commented:
A bit late, but a general comment: I am fine with the *functionality* being added to pandas, but I have some reservations about the need to have it as a DataFrame *method* (alternative would be a function, either top-level or in one of the submodules, eg testing). 

We already have many methods, and the barrier to add a new one should be relatively high. In this specific case you also have the overlap with the existing `diff`, and the two might be easy to confuse.

WillAyd commented:
Alternately we could get rid of the axis argument altogether with this. I think its ambiguous even along say the index if a user wants to see the diff between the frames "interwoven" or simply stacked one on top of another

Theres up to 4 ways to want the output aligned which one argument wouldn't cover, so I think could just pick one format and leave it to the user to reshape as desired

fujiaxiang committed:
added type indicator for method arguments

fujiaxiang commented:
I'm open with any suggestion. Indeed it is important that we discuss about the details and make a careful decision.

Here is my personal preferences / opinion:
1. I think this feature is primarily used as a convenience method during data exploration / pipeline testing / debugging phase, rather than used in production. Users may want to compare two different sets of outputs to decide what to do next, then discard the results. This is somewhat similar to the method `describe`. Hence, handiness is probably more important than how "complete" the arguments are
1. Implementing the functionality as a method would be the most convenience (i.e handy) for users. Top-level function is still ok, since most users would already have done `import pandas as pd`.
1. Regarding the confusion with existing method `diff`, perhaps we can consider renaming the method to `compare`?
1. Regarding @WillAyd's mention of "4 ways to align the output", I would disagree with this. Simply stacking the two objects together defeats the original intent of this function, which is to compare two data objects. Also, we already have `pd.concat` to achieve that. Again, this goes back to the value of "handiness". I would value a parameter to allow me to easily control the orientation of output alignment when I simply want to have a convenient way to compare two data pieces during data exploration. I would not want to take other steps to reshape this output since I just want to take a look at the result before discarding it.

In summary, I think we should keep it as a method, but we can consider renaming it to "compare".
Let me know what you guys think.

fujiaxiang committed:
updated type hints

fujiaxiang committed:
added NDFrame in FrameOrSeriesUnion type

simonjayhawkins commented on pandas/_typing.py:
```diff
@@ -52,7 +52,7 @@
 # is passed in, either a Series or DataFrame is returned, and if a DataFrame is passed
 # in, either a DataFrame or a Series is returned.
-FrameOrSeriesUnion = Union["DataFrame", "Series"]
+FrameOrSeriesUnion = Union["DataFrame", "NDFrame", "Series"]
```
I _think_ ommiting `NDFrame` from `FrameOrSeriesUnion` is intentional. can you revert this and replace `FrameOrSeriesUnion`  with `NDFrame` in the code where using `FrameOrSeriesUnion` is giving mypy errors.

fujiaxiang replied:
hi @simonjayhawkins, thanks a lot for review.

I'm struggling a bit to get mypy to work here. The problem I'm having is not with my own type hints, but the following line:
```python
diff = concat([self, other], axis=axis, keys=keys)
```
This gives me an error because `concat` accepts `FrameOrSeriesUnion`, but `self` has to be `NDFrame` or its parent class.
```python
def concat(
    objs: Union[Iterable[FrameOrSeriesUnion], Mapping[Label, FrameOrSeriesUnion]],
    ...
)
```

Do you have any suggestions how I should type hint `self` and `other` in my implementation in generic.py?

simonjayhawkins replied:
yeah, there looks to be an issue with concat type annotations. It's only the return type that should be `FrameOrSeriesUnion`, the arguments should be `FrameOrSeries` which includes `NDFrame`. this patch should fix.

```diff
diff --git a/pandas/_typing.py b/pandas/_typing.py
index 7387e100b..3b7392f78 100644
--- a/pandas/_typing.py
+++ b/pandas/_typing.py
@@ -52,7 +52,7 @@ FilePathOrBuffer = Union[str, Path, IO[AnyStr]]
 # `def func(a: FrameOrSeriesUnion) -> FrameOrSeriesUnion: ...` means that if a Series
 # is passed in, either a Series or DataFrame is returned, and if a DataFrame is passed
 # in, either a DataFrame or a Series is returned.
-FrameOrSeriesUnion = Union["DataFrame", "NDFrame", "Series"]
+FrameOrSeriesUnion = Union["DataFrame", "Series"]
 
 # FrameOrSeries is stricter and ensures that the same subclass of NDFrame always is
 # used. E.g. `def func(a: FrameOrSeries) -> FrameOrSeries: ...` means that if a
diff --git a/pandas/core/reshape/concat.py b/pandas/core/reshape/concat.py
index 091129707..eb195f836 100644
--- a/pandas/core/reshape/concat.py
+++ b/pandas/core/reshape/concat.py
@@ -6,7 +6,7 @@ from typing import Iterable, List, Mapping, Union, overload
 
 import numpy as np
 
-from pandas._typing import FrameOrSeriesUnion, Label
+from pandas._typing import FrameOrSeries, FrameOrSeriesUnion, Label
 
 from pandas.core.dtypes.generic import ABCDataFrame, ABCSeries
 
@@ -48,7 +48,7 @@ def concat(
 
 @overload
 def concat(
-    objs: Union[Iterable[FrameOrSeriesUnion], Mapping[Label, FrameOrSeriesUnion]],
+    objs: Union[Iterable[FrameOrSeries], Mapping[Label, FrameOrSeries]],
     axis=0,
     join: str = "outer",
     ignore_index: bool = False,
@@ -63,7 +63,7 @@ def concat(
 
 
 def concat(
-    objs: Union[Iterable[FrameOrSeriesUnion], Mapping[Label, FrameOrSeriesUnion]],
+    objs: Union[Iterable[FrameOrSeries], Mapping[Label, FrameOrSeries]],
     axis=0,
     join="outer",
     ignore_index: bool = False,
```

fujiaxiang replied:
Applied the patch and all mypy issues resolved.
Thanks a lot for this!

fujiaxiang committed:
fixed type hints of concat function

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

fujiaxiang commented:
guys, any thoughts?

simonjayhawkins commented:
generally lgtm ex @jorisvandenbossche comment https://github.com/pandas-dev/pandas/pull/30852#issuecomment-599128871

simonjayhawkins commented:
> In summary, I think we should keep it as a method, but we can consider renaming it to "compare".

+1 for compare

fujiaxiang committed:
renamed `differences` method to `compare`

fujiaxiang committed:
Merge branch 'master' into dataframe_and_series_differences

fujiaxiang committed:
correction of method name in doc/source/reference/series.rst

fujiaxiang commented:
@simonjayhawkins I have renamed the method to `compare`. Can you review again?

fujiaxiang committed:
added type checking in `compare` method and reformatted whatsnew a bit

fujiaxiang commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,31 @@ For example:
+
+.. ipython:: python
+
+   df
```
done, could you check?

fujiaxiang commented on pandas/core/generic.py:
```diff
@@ -8246,6 +8246,96 @@ def ranker(data):
+        keep_equal: bool_t = False,
+    ):
+        from pandas.core.reshape.concat import concat
+
```
added the type checking, thanks for suggestion

fujiaxiang commented on pandas/core/reshape/concat.py:
```diff
@@ -49,7 +49,7 @@ def concat(
 @overload
 def concat(
-    objs: Union[Iterable[FrameOrSeriesUnion], Mapping[Label, FrameOrSeriesUnion]],
+    objs: Union[Iterable[FrameOrSeries], Mapping[Label, FrameOrSeries]],
```
An earlier discussion concludes that the type hints here are incorrect: https://github.com/pandas-dev/pandas/pull/30852#discussion_r395765292

This makes it impossible for `compare` method to pass mypy checks with the type hints, so I am including this in the same PR.

fujiaxiang committed:
Merge branch 'master' into dataframe_and_series_differences

fujiaxiang committed:
removed unintended line break

fujiaxiang commented:
@jreback can you review?

fujiaxiang committed:
Merge branch 'master' into dataframe_and_series_differences

fujiaxiang committed:
resolved a linting issue

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

fujiaxiang committed:
updated whatsnew entry

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

fujiaxiang commented:
ping

fujiaxiang committed:
Merge branch 'master' into dataframe_and_series_differences

fujiaxiang commented:
ping

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -37,6 +37,37 @@ For example:
+Comparing two `DataFrame` or two `Series` and summarizing the differences
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We've added :meth:`DataFrame.compare` and :meth:`Series.compare` for comparing two `DataFrame` or two `Series` (:issue:`30429`)
```
can you add a similar example in in reshaping.rst or merge.rst (with a link from the other section) and a link here

fujiaxiang replied:
Added, can you review?

I have chosen to add the examples in merge.rst, as this `compare` method is more similar to a merging operation in that they both involved two aligned objects, and somehow combine them both.

jreback commented on pandas/core/generic.py:
```diff
@@ -8406,6 +8406,102 @@ def ranker(data):
+    _shared_docs[
+        "compare"
+    ] = """
+        Compare to another %(klass)s and show the differences.
```
can you add a vesionadded 1.1 tag here

fujiaxiang replied:
done

jreback commented on pandas/tests/frame/methods/test_compare.py:
```diff
@@ -0,0 +1,165 @@
+
+@pytest.mark.parametrize("align_axis", [0, 1, "index", "columns"])
+def test_compare_axis(align_axis):
+    df = pd.DataFrame(
```
can you add the issue number here as a comment

fujiaxiang replied:
done

jreback reviewed (commented):
looks good, a few small comments.

@pandas-dev/pandas-core any comments.

mroeschke commented:
Do we have a test where the two objects are not aligned?

e.g.

```
df1 = pd.DataFrame(np.ones((3,3)))
df2 = pd.DataFrame(np.zeros((2, 1)))
```

fujiaxiang committed:
added doc in user guide merging.rst and more tests

fujiaxiang commented:
> Do we have a test where the two objects are not aligned?
> 
> e.g.
> 
> ```
> df1 = pd.DataFrame(np.ones((3,3)))
> df2 = pd.DataFrame(np.zeros((2, 1)))
> ```

@mroeschke I have just added some. I'm expecting it to raise ``ValueError`` using `pytes.raises`.

fujiaxiang committed:
removed trailing space in docstring and blackified code

jreback commented on pandas/core/frame.py:
```diff
@@ -5689,6 +5689,108 @@ def _construct_result(self, result) -> "DataFrame":
+
+Keep all original rows and columns and also all original values
+
+>>> df.compare(df2, keep_shape=True, keep_equal=True)
```
does keep_equal make sense w/o keep_shape==True? IOW does it stand on its own? can you add an example of just using it

fujiaxiang replied:
Yes I would say so. It helps identify which one could be the "anomaly". I myself have come across such use cases. 

For example, say I'm looking at the date a company publishes 2 types of their quarterly report.
```python
>>> import pandas as pd
>>> df1 = pd.DataFrame(columns=['Q1_filing', 'Q2_filing', 'Q3_filing', 'Q4_filing'], index=list(range(2010, 2020)))
>>> df1['Q1_filing'] = 'Mar'
>>> df1['Q2_filing'] = 'Jun'
>>> df1['Q3_filing'] = 'Sep'
>>> df1['Q4_filing'] = 'Dec'
>>> df1
     Q1_filing Q2_filing Q3_filing Q4_filing
2010       Mar       Jun       Sep       Dec
2011       Mar       Jun       Sep       Dec
2012       Mar       Jun       Sep       Dec
2013       Mar       Jun       Sep       Dec
2014       Mar       Jun       Sep       Dec
2015       Mar       Jun       Sep       Dec
2016       Mar       Jun       Sep       Dec
2017       Mar       Jun       Sep       Dec
2018       Mar       Jun       Sep       Dec
2019       Mar       Jun       Sep       Dec

>>> df2 = df1.copy()
>>> df2.loc[2015, 'Q1_filing'] = 'Apr'
>>> df2.loc[2016, 'Q2_filing'] = 'Jul'
```

By comparing the two, I can see that the discrepancy is in 2015 and 2016, but I don't know which one deviated from the norm.
```python
>>> df1.compare(df2)
     Q1_filing       Q2_filing
          self other      self other
2015       Mar   Apr       NaN   NaN
2016       NaN   NaN       Jun   Jul
```

The natural thing for me to do now is look at 2015 Q2_filing and 2016 Q1_filing where they agree with each other. (You can of course look at the whole thing but sometimes data is too big and I just want to take a look at the relevant ones first)
```python
>>> df1.compare(df2, keep_equal=True)
     Q1_filing       Q2_filing
          self other      self other
2015       Mar   Apr       Jun   Jun
2016       Mar   Mar       Jun   Jul
```
With this result I know probably something is off for the second type of reports.

fujiaxiang replied:
Have added the example in frame.py. ``keep_equal`` does not stand on its own for `Series` so I did not add anything there.

jreback commented:
question above and can you merge master.

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into dataframe_and_series_differences

# Conflicts:
#	pandas/core/series.py

fujiaxiang committed:
added one more example in docstring of DataFrame.compare

fujiaxiang commented:
> question above and can you merge master.

answered above

jreback commented:
k thanks @fujiaxiang 

any other comments @pandas-dev/pandas-core  will merge in 24 hours otherwise.

WillAyd reviewed (approved):
lgtm

jreback commented:
thanks @fujiaxiang very nice!
